README.md
# 🎖 Installing

1) First goto your pubspec.yaml file and add dependency:
   ```
   flutter_network_service:
   ```
2) run flutter pub get to update the packages dependencies.

# ⚡️ Import
3) In your code, you can use it like this:
  first import the package:
  ```
 import 'package:flutter_network_service/HttpNetworkService.dart';
  ```
# 🎮 Code Usage
then use it like this in your code:
  ```

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutternetworkservicehandler/src/http_network_service.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List _usersList = [];

  var errorMessage = 'Can\'t proceed request';

  void _showErrorDialog(String message) {
    showDialog(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Error Occured'),
        content: Text(message),
        actions: <Widget>[
          ElevatedButton(
            onPressed: () {
              Navigator.of(ctx).pop();
            },
            child: Text('OK!'),
          )
        ],
      ),
    );
  }

  Future getUsersData() async {
    try {
      //Initialize HttpNetwork Service
      await HttpNetworkService.init(
        baseUrl: 'https://jsonplaceholder.typicode.com',
      );
      //Then use it like this
      final responseUsersList = await HttpNetworkService.getRequest(
        uri: '/userss',
        //options import from dio package.
        options: Options(headers: {
          "Content-Type": "application/json",
        }),
      );
      for (var items in responseUsersList) {
        print(items);
        _usersList.add({
          'id': items['id'],
          'username': items['username'],
          'website': items['website'],
        });
      }
    } catch (error) {
      errorMessage = error.toString();
      _showErrorDialog(
        errorMessage,
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FutureBuilder(
        future: getUsersData(),
        builder: (ctx, snapshot) =>
            snapshot.connectionState == ConnectionState.waiting
                ? Center(
                    child: CircularProgressIndicator(),
                  )
                : snapshot.data == null
                    ? Center(
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Icon(
                              Icons.error,
                              size: 75,
                              color: Colors.red,
                            ),
                            Padding(
                              padding: const EdgeInsets.all(8.0),
                              child: Text(
                                errorMessage.toUpperCase(),
                                textAlign: TextAlign.center,
                                style: TextStyle(
                                  fontSize: 16.0,
                                  fontWeight: FontWeight.w700,
                                ),
                              ),
                            ),
                          ],
                        ),
                      )
                    : ListView.builder(
                        itemCount: _usersList.length,
                        itemBuilder: (ctx, i) => ListTile(
                          title: Text(_usersList[i]['username']),
                          subtitle: Text(_usersList[i]['website']),
                          leading: CircleAvatar(
                            child: FittedBox(
                              child: Text(
                                _usersList[i]['id'].toString(),
                              ),
                            ),
                          ),
                          trailing: Icon(
                            Icons.arrow_forward_ios,
                          ),
                        ),
                      ),
      ),
    );
  }
}

```
# FOR MORE EXAMPLES REFER TO example/lib folder.

# 🐛 Bugs/Requests
If you encounter any problems feel free to open an issue. If you feel the library is
missing some feature, please raise a ticket on Github. Pull request are also welcome.

# 🚀 SUPPORT
You liked this package? then give it a star. If you want to help then:
Star this repository.
Send a Pull Request with new features, see Contribution Guidelines.
Share this package.
Create issues if you find a Bug or want to suggest something.



CHANGELOG.md
[0.0.1] - 23-Dec-2020
Initial release of flutter network service handler

[0.0.2] - 24-Dec-2020
Added Status code handling with http exception custom class on flutter network service handler

[0.0.3] - 08-Jan-2021
Added Internet Connectivity check.

[0.0.4] - 08-Jan-2021
Added Internet Connectivity check using data connection checker.

[0.0.5] - 08-Jan-2021
Replace data connection checker with connectivity package to handle Internet check for all the platforms.

[0.0.6] - 06-Jan-2021
Replace http with dio, added image uploading support etc.


lib/src/dio_client.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

import 'http_exception.dart';

const _defaultConnectTimeout = Duration.millisecondsPerMinute;
const _defaultReceiveTimeout = Duration.millisecondsPerMinute;
const kServerError =
    'The server encountered an internal error and unable to process your request.';
const kRedirectionError = 'The resource requested has been temporarily moved.';
const kBadRequestError =
    'Your client has issued a malformed or illegal request.';
const kInternetError =
    'There is poor or no internet connection, please try again later.';

class DioClient {
  Dio? _dio;
  String? baseUrl;

  final List<Interceptor>? interceptors;

  DioClient({
    required Dio dio,
    this.interceptors,
    required this.baseUrl,
  }) {
    _dio = dio;
    _dio!
      ..options.baseUrl = baseUrl!
      ..options.connectTimeout = _defaultConnectTimeout
      ..options.receiveTimeout = _defaultReceiveTimeout
      ..options.headers = {'Content-Type': 'application/json; charset=UTF-8'}
      ..httpClientAdapter;

    if (interceptors?.isNotEmpty ?? false) {
      _dio!.interceptors.addAll(interceptors!);
    }
    if (kDebugMode) {
      _dio!.interceptors.add(LogInterceptor(
          responseBody: true,
          error: true,
          requestHeader: true,
          responseHeader: true,
          request: true,
          requestBody: true));
    }
  }

  Future<dynamic> get(
    String? uri, {
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    ProgressCallback? onReceiveProgress,
  }) async {
    try {
      var response = await _dio!.get(
        uri!,
        queryParameters: queryParameters,
        options: options,
        cancelToken: cancelToken,
        onReceiveProgress: onReceiveProgress,
      );
      return response.data;
    } on DioError catch (e) {
      if (DioErrorType.receiveTimeout == e.type ||
          DioErrorType.connectTimeout == e.type ||
          DioErrorType.sendTimeout == e.type) {
        throw HttpException(kInternetError);
      } else if (DioErrorType.response == e.type) {
        //when api returns error then use this.

        if (e.response!.statusCode! >= 400 || e.response!.statusCode! <= 499) {
          throw HttpException(kBadRequestError);
        } else if (e.response!.statusCode! >= 300 ||
            e.response!.statusCode! <= 399) {
          throw HttpException(kRedirectionError);
        } else if (e.response!.statusCode! >= 500 ||
            e.response!.statusCode! <= 599) {
          throw HttpException(kServerError);
        }
      } else {
        throw HttpException(kServerError);
      }
    }
  }

  Future<dynamic> delete(
    String? uri, {
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
  }) async {
    try {
      var response = await _dio!.delete(
        uri!,
        queryParameters: queryParameters,
        options: options,
        cancelToken: cancelToken,
      );

      return response.data;
    } on DioError catch (e) {
      if (DioErrorType.receiveTimeout == e.type ||
          DioErrorType.connectTimeout == e.type ||
          DioErrorType.sendTimeout == e.type) {
        throw HttpException(kInternetError);
      } else if (DioErrorType.response == e.type) {
        //when api returns error then use this.

        if (e.response!.statusCode! >= 400 || e.response!.statusCode! <= 499) {
          throw HttpException(kBadRequestError);
        } else if (e.response!.statusCode! >= 300 ||
            e.response!.statusCode! <= 399) {
          throw HttpException(kRedirectionError);
        } else if (e.response!.statusCode! >= 500 ||
            e.response!.statusCode! <= 599) {
          throw HttpException(kServerError);
        }
      } else {
        throw HttpException(kServerError);
      }
    }
  }

  Future<dynamic> post(
    String? uri, {
    data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) async {
    try {
      var response = await _dio!.post(
        uri!,
        data: data,
        queryParameters: queryParameters,
        options: options,
        cancelToken: cancelToken,
        onSendProgress: onSendProgress,
        onReceiveProgress: onReceiveProgress,
      );

      return response.data;
    } on DioError catch (e) {
      if (DioErrorType.receiveTimeout == e.type ||
          DioErrorType.connectTimeout == e.type ||
          DioErrorType.sendTimeout == e.type) {
        throw HttpException(kInternetError);
      } else if (DioErrorType.response == e.type) {
        //when api returns error then use this.

        if (e.response!.statusCode! >= 400 || e.response!.statusCode! <= 499) {
          throw HttpException(kBadRequestError);
        } else if (e.response!.statusCode! >= 300 ||
            e.response!.statusCode! <= 399) {
          throw HttpException(kRedirectionError);
        } else if (e.response!.statusCode! >= 500 ||
            e.response!.statusCode! <= 599) {
          throw HttpException(kServerError);
        }
      } else {
        throw HttpException(kServerError);
      }
    }
  }
}


example/lib/example.dart
import 'package:flutter/material.dart';
import 'package:flutternetworkservicehandler/src/http_network_service.dart';
import 'package:flutternetworkservicehandler/src/http_exception.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List _usersList = [];

  var errorMessage = 'Can\'t proceed request';

  void _showErrorDialog(String message) {
    showDialog(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Error Occured'),
        content: Text(message),
        actions: <Widget>[
          FlatButton(
            onPressed: () {
              Navigator.of(ctx).pop();
            },
            child: Text('OK!'),
          )
        ],
      ),
    );
  }

  Future getUsersData() async {
    try {
      final responseUsersList = await HttpNetworkService.getRequest(
        url: 'https://jsonplaceholder.typicode.com/userss',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      );
      for (var items in responseUsersList) {
        print(items);
        _usersList.add({
          'id': items['id'],
          'username': items['username'],
          'website': items['website'],
        });
      }
    } on HttpException catch (error) {
      if (error.toString().contains('Redirection error')) {
        errorMessage = 'The resource requested has been temporarily moved.';
      } else if (error.toString().contains('Bad Request Format')) {
        errorMessage = 'Your client has issued a malformed or illegal request.';
      } else if (error.toString().contains('Internal Server Error')) {
        errorMessage =
            'The server encountered an error and could not complete your request.';
      } else if (error.toString().contains('No Internet Found')) {
        errorMessage =
            'There is no or poor internet connect. Please try again later';
      }
      _showErrorDialog(errorMessage);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FutureBuilder(
        future: getUsersData(),
        builder: (ctx, snapshot) =>
            snapshot.connectionState == ConnectionState.waiting
                ? Center(
                    child: CircularProgressIndicator(),
                  )
                : snapshot.data == null
                    ? Center(
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Icon(
                              Icons.error,
                              size: 75,
                              color: Colors.red,
                            ),
                            Padding(
                              padding: const EdgeInsets.all(8.0),
                              child: Text(
                                errorMessage.toUpperCase(),
                                textAlign: TextAlign.center,
                                style: TextStyle(
                                  fontSize: 16.0,
                                  fontWeight: FontWeight.w700,
                                ),
                              ),
                            ),
                          ],
                        ),
                      )
                    : ListView.builder(
                        itemCount: _usersList.length,
                        itemBuilder: (ctx, i) => ListTile(
                          title: Text(_usersList[i]['username']),
                          subtitle: Text(_usersList[i]['website']),
                          leading: CircleAvatar(
                            child: FittedBox(
                              child: Text(
                                _usersList[i]['id'].toString(),
                              ),
                            ),
                          ),
                          trailing: Icon(
                            Icons.arrow_forward_ios,
                          ),
                        ),
                      ),
      ),
    );
  }
}


lib/src/http_network_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:flutter_network_service/src/dio_client.dart';

import 'http_exception.dart';

class HttpNetworkService {
  static late DioClient _dioClient;

  static init({
    required String? baseUrl,
  }) {
    var dio = Dio();

    _dioClient = DioClient(
      baseUrl: baseUrl,
      dio: dio,
    );
  }

  //it sends a get request using dio package with exception handling
  static Future getRequest(
      {String? uri,
      Map<String, dynamic>? queryParameters,
      Options? options,
      CancelToken? cancelToken,
      ProgressCallback? onReceiveProgress}) async {
    try {
      final response = await _dioClient.get(
        uri,
        queryParameters: queryParameters,
        cancelToken: cancelToken,
        onReceiveProgress: onReceiveProgress,
        options: options,
      );

      return response;
    } on HttpException catch (e) {
      throw e.message;
    }
  }

  //it sends a post request using dio package with exception handling
  static Future postRequest({
    String? uri,
    data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) async {
    try {
      final response = await _dioClient.post(
        uri,
        options: options,
        onReceiveProgress: onReceiveProgress,
        cancelToken: cancelToken,
        queryParameters: queryParameters,
        data: data,
        onSendProgress: onSendProgress,
      );
      return response;
    } on HttpException catch (e) {
      throw e.message;
    }
  }

  //it sends a delete request using dio package with exception handling
  static Future deleteRequest(
      {String? uri,
      Map<String, dynamic>? queryParameters,
      Options? options,
      CancelToken? cancelToken}) async {
    try {
      final response = await _dioClient.delete(
        uri,
        queryParameters: queryParameters,
        cancelToken: cancelToken,
        options: options,
      );
      return response;
    } on HttpException catch (e) {
      throw e.message;
    }
  }

  //Send single or multiple images to server using dio
  static Future sendImagesRequest(
      {required String? uri,
      Options? options,
      CancelToken? cancelToken,
      ProgressCallback? onSendProgress,
      ProgressCallback? onReceiveProgress,
      Map<String, dynamic>? queryParameters,
      required List<File> images,
      required String? fileName}) async {
    try {
      var formData = FormData();

      images.forEach((image) async {
        formData.files.add(MapEntry(
          fileName!,
          await MultipartFile.fromFile(
            '${image.path}',
            filename: fileName,
          ),
        ));
      });

      final response = await _dioClient.post(
        uri,
        data: formData,
        options: options,
        cancelToken: cancelToken,
        queryParameters: queryParameters,
        onSendProgress: onSendProgress,
        onReceiveProgress: onReceiveProgress,
      );
      return response;
    } on HttpException catch (e) {
      throw e.message;
    }
  }
}


example/lib/send_image_to_server.dart
/// This contains the basic understanding of how to send a single or multiple images to server, you need to pass the list of files
/// of images
/// You can integrate it easily with your flutter apps.
/// Api throws httpexception which is handled accordingly just use simple try and catch block and everything will be
/// handled by the package


import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter_network_service/src/http_network_service.dart';



void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List _usersList = [];

  var errorMessage = '';

  void _showErrorDialog(String message) {
    showDialog(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Error Occured'),
        content: Text(message),
        actions: <Widget>[
          ElevatedButton(
            onPressed: () {
              Navigator.of(ctx).pop();
            },
            child: Text('OK!'),
          )
        ],
      ),
    );
  }

  Future sendImageToServer() async {
    try {
      //Initialize HttpNetwork Service
      await HttpNetworkService.init(
        baseUrl: 'localhost',
      );
      //Then use it like this
      final response = await HttpNetworkService.sendImagesRequest(
        uri: '/blog-post',
        fileName: "BlogImg",
        images: [File("img1.png"), File("img2.png"), File("img3.png")],
        //options import from dio package.
        options: Options(headers: {
          "Content-Type": "application/json",
        }),
      );
      if (response != null) {
        debugPrint("Images send success");
      }
    } catch (error) {
      errorMessage = error.toString();
      _showErrorDialog(
        errorMessage,
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: SafeArea(
      child: Center(
        child: errorMessage.isEmpty
            ? ElevatedButton(
                onPressed: () async {
                  await sendImageToServer();
                },
                child: const Text(
                  "Send image to server",
                ),
              )
            : Padding(
                padding: const EdgeInsets.all(8.0),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    const Text(
                      "Send image to server",
                    ),
                    const SizedBox(
                      height: 10,
                    ),
                    ElevatedButton(
                      onPressed: () async {
                        await sendImageToServer();
                      },
                      child: const Text(
                        "Send image to server Again",
                      ),
                    )
                  ],
                ),
              ),
      ),
    ));
  }
}


example/lib/basic_example.dart
/// THIS CONTAINS THE BASIC GET REQUEST USING flutternetworkservicehandler
/// You can integrate it easily with your flutter apps.
/// Api throws httpexception which is handled accordingly just use simple try and catch block and everything will be
/// handled by the package

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter_network_service/flutter_network_service.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List _usersList = [];

  var errorMessage = 'Can\'t proceed request';

  void _showErrorDialog(String message) {
    showDialog(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Error Occured'),
        content: Text(message),
        actions: <Widget>[
          ElevatedButton(
            onPressed: () {
              Navigator.of(ctx).pop();
            },
            child: Text('OK!'),
          )
        ],
      ),
    );
  }

  Future getUsersData() async {
    try {
      //Initialize HttpNetwork Service
      await HttpNetworkService.init(
        baseUrl: 'https://jsonplaceholder.typicode.com',
      );
      //Then use it like this
      final responseUsersList = await HttpNetworkService.getRequest(
        uri: '/userss',
        //options import from dio package.
        options: Options(headers: {
          "Content-Type": "application/json",
        }),
      );
      for (var items in responseUsersList) {
        print(items);
        _usersList.add({
          'id': items['id'],
          'username': items['username'],
          'website': items['website'],
        });
      }
    } catch (error) {
      errorMessage = error.toString();
      _showErrorDialog(
        errorMessage,
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FutureBuilder(
        future: getUsersData(),
        builder: (ctx, snapshot) =>
            snapshot.connectionState == ConnectionState.waiting
                ? Center(
                    child: CircularProgressIndicator(),
                  )
                : snapshot.data == null
                    ? Center(
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Icon(
                              Icons.error,
                              size: 75,
                              color: Colors.red,
                            ),
                            Padding(
                              padding: const EdgeInsets.all(8.0),
                              child: Text(
                                errorMessage.toUpperCase(),
                                textAlign: TextAlign.center,
                                style: TextStyle(
                                  fontSize: 16.0,
                                  fontWeight: FontWeight.w700,
                                ),
                              ),
                            ),
                          ],
                        ),
                      )
                    : ListView.builder(
                        itemCount: _usersList.length,
                        itemBuilder: (ctx, i) => ListTile(
                          title: Text(_usersList[i]['username']),
                          subtitle: Text(_usersList[i]['website']),
                          leading: CircleAvatar(
                            child: FittedBox(
                              child: Text(
                                _usersList[i]['id'].toString(),
                              ),
                            ),
                          ),
                          trailing: Icon(
                            Icons.arrow_forward_ios,
                          ),
                        ),
                      ),
      ),
    );
  }
}
